lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -405,342 +861,1506 @@ bool ADContext::supportsVectorDifferentiation(Type type) const {
+    builder.createAllocStack(loc, typeConv.getLoweredType(scalarTy));
+  SWIFT_DEFER {
+    // dealloc_stack %2 : $<scalar type>
+    builder.createDeallocStack(loc, scalarValBuf);
+  };
+  // store %0 : $<scalar type> to $*<scalar type>
+  builder.createStore(loc, scalarVal, scalarValBuf,
+                      getBufferSOQ(scalarTy, module));
+  auto *vecNumProto = context.getVectorNumericProtocol();
+  auto *reqr =
+    cast<ConstructorDecl>(vecNumProto->lookupDirect(initName).front());
+  SILDeclRef reqrRef(reqr, SILDeclRef::Kind::Allocator);
+  auto silInitTy = context.getTypeConverter().getConstantType(reqrRef);
+  // Get scalar's conformance to `FloatingPoint`.
+  auto conf = astCtx.getConformance(type, vecNumProto,
+                                    targetTypeDecl->getLoc(), targetTypeDecl,
+                                    ProtocolConformanceState::Complete);
+  ProtocolConformanceRef confRef(conf);
+  // $4 = witness_method ...
+  auto initFnRef =
+    builder.createWitnessMethod(loc, type, confRef, reqrRef, silInitTy);
+  auto initSubMap =
+    SubstitutionMap::getProtocolSubstitutions(vecNumProto, type, confRef);
+  // %5 = apply %4(%3, %2, %1)
+  builder.createApply(loc, initFnRef, initSubMap,
+                      { seedBuf, scalarValBuf, metatype },
+                      /*isNonThrowing*/ false);
+}
+
+//===----------------------------------------------------------------------===//
+// PrimalGen - generates primal functions for each differentiation task in
+// the SIL module.
+//===----------------------------------------------------------------------===//
+
+namespace {
+class PrimalGen {
+  friend class PrimalGenCloner;
+private:
+  /// The global AD context.
+  ADContext &context;
+  /// A worklist of primal synthesis items, each of which specifies a the
+  /// original function, the target primal function, AD indices, and the primal
+  /// value struct.
+  SmallVector<FunctionSynthesisItem, 16> worklist;
+
+public:
+  explicit PrimalGen(ADContext &context) : context(context) {}
+
+  void run();
+
+protected:
+  /// Lazily create a task to synthesize the primal function.
+  SILFunction *lookupPrimalOrScheduleSynthesis(DifferentiationTask *task);
+
+private:
+  /// Creates an empty primal function, updating the primal info in the task.
+  std::pair<SILFunction *, StructDecl *>
+  createEmptyPrimal(DifferentiationTask *task);
+
+  /// Processes an original function and generate its adjoint.
+  void performSynthesis(FunctionSynthesisItem task);
+};
+} // end anonymous namespace
+
+/// Creates a struct declaration (without contents) for storing contextual
+/// information for a function. The newly created struct will have the same
+/// generic parameters as the function.
+///
+/// This helper function will be used to create checkpointing structs,
+/// specifically structs for control-independent checkpoints and tape groups.
+StructDecl *ADContext::createDependentStructForFunction(SILFunction *function) {
```

Comment:
Is there a reason for the difference in terminology between "dependent struct" and "primal value struct"?